File: regions/system/library/typescript/auth/src/pkce.ts
/**
 * PKCE (Proof Key for Code Exchange) ユーティリティ
 * RFC 7636 準拠
 */

/**
 * ランダムな code_verifier を生成する。
 * Base64url エンコードされた 32 バイトのランダム値。
 * @param getRandomValues - crypto.getRandomValues 互換の関数（テスト用に注入可能）
 */
export function generateCodeVerifier(
  getRandomValues: (array: Uint8Array) => Uint8Array = (array) =>
    crypto.getRandomValues(array),
): string {
  const array = new Uint8Array(32);
  getRandomValues(array);
  return base64UrlEncode(array);
}

/**
 * code_verifier から code_challenge を計算する（S256）。
 * SHA-256 ハッシュの Base64url エンコード。
 * @param codeVerifier - PKCE code_verifier
 * @param subtleDigest - crypto.subtle.digest 互換の関数（テスト用に注入可能）
 */
export async function generateCodeChallenge(
  codeVerifier: string,
  subtleDigest: (
    algorithm: string,
    data: BufferSource,
  ) => Promise<ArrayBuffer> = (algorithm, data) =>
    crypto.subtle.digest(algorithm, data),
): Promise<string> {
  const encoder = new TextEncoder();
  const data = encoder.encode(codeVerifier);
  const digest = await subtleDigest('SHA-256', data);
  return base64UrlEncode(new Uint8Array(digest));
}

/**
 * Uint8Array を Base64url エンコードする。
 * RFC 4648 Section 5 準拠（パディングなし）。
 */
export function base64UrlEncode(buffer: Uint8Array): string {
  let binary = '';
  for (let i = 0; i < buffer.length; i++) {
    binary += String.fromCharCode(buffer[i]);
  }
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

File: regions/system/library/dart/auth/lib/src/pkce.dart
/// PKCE (Proof Key for Code Exchange) ユーティリティ
/// RFC 7636 準拠
library;

import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';

/// ランダムな code_verifier を生成する。
/// RFC 7636 Section 4.1 準拠。
/// [random] パラメータはテスト用に注入可能。
String generateCodeVerifier({Random? random}) {
  final rng = random ?? Random.secure();
  const chars =
      'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
  return List.generate(43, (_) => chars[rng.nextInt(chars.length)]).join();
}

/// code_verifier から code_challenge を計算する（S256）。
/// SHA-256 ハッシュの Base64url エンコード（パディングなし）。
String generateCodeChallenge(String codeVerifier) {
  final bytes = utf8.encode(codeVerifier);
  final digest = sha256.convert(bytes);
  return base64UrlEncode(digest.bytes);
}

/// バイト列を Base64url エンコードする（パディングなし）。
/// RFC 4648 Section 5 準拠。
String base64UrlEncode(List<int> bytes) {
  return base64Url.encode(bytes).replaceAll('=', '');
}
